File: ant_books/2_1/dfs.rs
fn dfs(i: usize, sum: usize, t: usize, a: &[usize]) -> bool {
  if i == a.len() { return sum == t };
  if dfs(i+1, sum, t, a) { return true };
  if dfs(i+1, sum + a[i], t, a) { return true };
  false
}

fn main() {
  let a = [1, 2, 3, 4];
  let t = 7;
  println!("Make {} in {:?}", t, a);
  println!("{}", if dfs(0, 0, t, &a) { "Yes" } else { "No" });
}

File: ant_book/2_1/dfs.rb
A = [1, 2, 4, 7]
T = 15
STACK = []

def search
  sum = STACK.sum
  if sum == T
    p STACK
    return true
  end
  return false if sum > T

  for i in 0..(A.length-1) do
    next unless A[i]
    STACK.push(A[i])
    A[i] = nil

    if search
      return true
    else
      A[i] = STACK.pop
    end
  end
  false
end

puts "Search #{T}"
puts search ? "Yes" : "No"
